File: algo/src/BOJ/Silver/BOJ1629.java
package BOJ.Silver;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class BOJ1629 {//풀이중
    public static void main(String[] args) throws IOException {
        BufferedReader br =new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        long A =Integer.parseInt(st.nextToken());
        long B = Integer.parseInt(st.nextToken());
        long C = Integer.parseInt(st.nextToken());
        long sum= 1;

        System.out.println((int)(Math.pow(A,B)%C));
    }
}

File: PS/algo/src/BOJ/Silver/BOJ1629.kt
package BOJ.silver

fun main() {
    val br = System.`in`.bufferedReader()
    val (a, b, c) = br.readLine().split(" ").map { it.toLong() }
    println(find(a, b, c,).mod(c))
}

private fun find(a:Long,b:Long,c:Long):Long{
    if(b==1L) return a % c

    val temp = find(a,b/2,c)

    return if(b % 2 ==0L) (temp * temp ).mod(c) else ((temp * temp ).mod(c) * a).mod(c)
}
